Rename _scatter_reduce to scatter_reduce

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #71256
* __->__ #70750

- Change name from `_scatter_reduce` to `scatter_reduce`
- Make `scatter_reduce` unstructured because gradients not supported for out

Differential Revision: [D33712801](https://our.internmc.facebook.com/intern/diff/D33712801)

mikaylagawarecki committed:
Rename _sparse_tensor sparse_tensor2

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Rename _sparse_tensor sparse_tensor2"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Rename _sparse_tensor sparse_tensor2"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Rename _sparse_tensor sparse_tensor"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Rename _sparse_tensor sparse_tensor"




[ghstack-poisoned]

cpuhrsch commented:
@pytorchbot ciflow rerun

mikaylagawarecki committed:
Update on "Rename _sparse_tensor sparse_tensor"




[ghstack-poisoned]

mikaylagawarecki commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -1299,16 +1277,41 @@ TORCH_IMPL_FUNC(scatter_add)
+                              const Tensor& index,
+                              const c10::string_view reduce,
+                              const c10::optional<int64_t> output_size,
+                              const c10::optional<Tensor>& optional_out) {
```
just want to double check that I didn't misunderstand - by making `scatter_reduce` unstructured and removing the out variant that means we still want the optional_out argument from `torch_scatter` so gradients can be calculated, is that correct? If not I'll remove this additional argument and undo the changes!

cpuhrsch replied:
Let's remove this argument for now and introduce it later on. Our common definition of out variants doesn't support gradients and we shouldn't let this one-off operator intersect with that.

mikaylagawarecki committed:
Update on "Rename _scatter_reduce to scatter_reduce"


- Change name from `_scatter_reduce` to `scatter_reduce`
- Make `scatter_reduce` unstructured so `optional_out` argument can have gradients for now

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712801).

albanD reviewed (commented):
This is missing proper documentation etc as you're making it a public function?

cpuhrsch commented:
@albanD - docs will land as part of https://github.com/pytorch/pytorch/pull/68580

This PR unblocks that.

mikaylagawarecki committed:
Update on "Rename _scatter_reduce to scatter_reduce"


- Change name from `_scatter_reduce` to `scatter_reduce`
- Make `scatter_reduce` unstructured because gradients not supported for out

Differential Revision: [D33712801](https://our.internmc.facebook.com/intern/diff/D33712801)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712801).

albanD commented:
Note that the doc update will need to land before the branch cut. Or we will have to remove this function from the public API for the release.
